vyper/parser/expr.py

```diff
@@ -521,117 +505,95 @@ def arithmetic(self):
-                    pos=getpos(self.expr),
-                )
+                # TODO should this be 'smul' (note edge cases in YP for smul)
+                arith = ['mul', 'l', 'r']
+
             elif ltyp == rtyp == 'decimal':
-                o = LLLnode.from_list([
-                    'with', 'r', right, [
-                        'with', 'l', left, [
-                            'with', 'ans', ['mul', 'l', 'r'],
-                            [
-                                'seq',
-                                [
-                                    'assert',
-                                    ['or', ['eq', ['sdiv', 'ans', 'l'], 'r'], ['iszero', 'l']]
-                                ],
-                                ['sdiv', 'ans', DECIMAL_DIVISOR],
-                            ],
-                        ],
-                    ],
-                ], typ=BaseType('decimal', new_unit), pos=getpos(self.expr))
+                arith = ['with', 'ans', ['mul', 'l', 'r'],
+                         ['seq',
+                             ['assert',
+                                 ['or',
+                                     ['eq', ['sdiv', 'ans', 'l'], 'r'],
+                                     ['iszero', 'l']]],
+                             ['sdiv', 'ans', DECIMAL_DIVISOR]]]
             else:
                 raise Exception(f"Unsupported Operation 'mul({ltyp}, {rtyp})'")
+
         elif isinstance(self.expr.op, ast.Div):
             if left.typ.positional or right.typ.positional:
                 raise TypeMismatchException("Cannot divide positional values!", self.expr)
             new_unit = combine_units(left.typ.unit, right.typ.unit, div=True)
+            new_typ = BaseType(ltyp, new_unit)
             if ltyp == rtyp == 'uint256':
-                o = LLLnode.from_list([
-                    'seq',
-                    # Checks that:  b != 0
-                    ['assert', right],
-                    ['div', left, right],
-                ], typ=BaseType('uint256', new_unit), pos=getpos(self.expr))
+                arith = ['div', 'l', ['clamp_nonzero', 'r']]
+
             elif ltyp == rtyp == 'int128':
-                o = LLLnode.from_list(
-                    ['sdiv', left, ['clamp_nonzero', right]],
-                    typ=BaseType('int128', new_unit),
-                    pos=getpos(self.expr),
-                )
+                arith = ['sdiv', 'l', ['clamp_nonzero', 'r']]
+
             elif ltyp == rtyp == 'decimal':
-                o = LLLnode.from_list([
-                    'with', 'l', left, [
-                        'with', 'r', ['clamp_nonzero', right], [
-                            'sdiv',
-                            ['mul', 'l', DECIMAL_DIVISOR],
-                            'r',
-                        ],
-                    ]
-                ], typ=BaseType('decimal', new_unit), pos=getpos(self.expr))
+                arith = ['sdiv',
+                         # TODO check overflow cases
```

Comment:
Let's fix this TODO

Reply:
i think it's covered by the clamp later actually